// appium/build/test/helpers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.W3C_PREFIXED_CAPS = exports.W3C_CAPS = exports.TEST_HOST = exports.TEST_FAKE_APP = exports.PROJECT_ROOT = exports.BASE_CAPS = void 0;
exports.getTestPort = getTestPort;
exports.resolveFixture = resolveFixture;
Object.defineProperty(exports, "rewiremock", {
  enumerable: true,
  get: function () {
    return _rewiremock.default;
  }
});

require("source-map-support/register");

var _rewiremock = _interopRequireWildcard(require("rewiremock"));

var _path = _interopRequireDefault(require("path"));

var _utils = require("../lib/utils");

var _getPort = _interopRequireDefault(require("get-port"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const TEST_HOST = '127.0.0.1';
exports.TEST_HOST = TEST_HOST;

const fakeDriverPath = _path.default.dirname(require.resolve('@appium/fake-driver/package.json'));

const PROJECT_ROOT = _path.default.join(fakeDriverPath, '..', '..');

exports.PROJECT_ROOT = PROJECT_ROOT;

const TEST_FAKE_APP = _path.default.join(fakeDriverPath, 'test', 'fixtures', 'app.xml');

exports.TEST_FAKE_APP = TEST_FAKE_APP;
const BASE_CAPS = {
  automationName: 'Fake',
  platformName: 'Fake',
  deviceName: 'Fake',
  app: TEST_FAKE_APP
};
exports.BASE_CAPS = BASE_CAPS;
const W3C_PREFIXED_CAPS = { ...(0, _utils.insertAppiumPrefixes)(BASE_CAPS)
};
exports.W3C_PREFIXED_CAPS = W3C_PREFIXED_CAPS;
const W3C_CAPS = {
  alwaysMatch: { ...W3C_PREFIXED_CAPS
  },
  firstMatch: [{}]
};
exports.W3C_CAPS = W3C_CAPS;
let TEST_PORT;

async function getTestPort() {
  return await (TEST_PORT || (0, _getPort.default)());
}

function resolveFixture(filename, ...pathParts) {
  return _path.default.join(__dirname, 'fixtures', filename, ...pathParts);
}

(0, _rewiremock.overrideEntryPoint)(module);
(0, _rewiremock.addPlugin)(_rewiremock.plugins.nodejs);
(0, _rewiremock.addPlugin)(_rewiremock.plugins.childOnly);require('source-map-support').install();
