// @appium/support/build/test/logger/helpers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.assertOutputContains = assertOutputContains;
exports.assertOutputDoesntContain = assertOutputDoesntContain;
exports.getDynamicLogger = getDynamicLogger;
exports.restoreWriters = restoreWriters;
exports.setupWriters = setupWriters;

require("source-map-support/register");

var _sinon = _interopRequireDefault(require("sinon"));

var _lodash = _interopRequireDefault(require("lodash"));

var _lib = require("../../lib");

function setupWriters() {
  return {
    'stdout': _sinon.default.spy(process.stdout, 'write'),
    'stderr': _sinon.default.spy(process.stderr, 'write')
  };
}

function getDynamicLogger(testingMode, forceLogs, prefix = null) {
  process.env._TESTING = testingMode ? '1' : '0';
  process.env._FORCE_LOGS = forceLogs ? '1' : '0';
  return _lib.logger.getLogger(prefix);
}

function restoreWriters(writers) {
  for (let w of _lodash.default.values(writers)) {
    w.restore();
  }
}

function someoneHadOutput(writers, output) {
  let hadOutput = false;

  let matchOutput = _sinon.default.match(function (value) {
    return value && value.indexOf(output) >= 0;
  }, 'matchOutput');

  for (let writer of _lodash.default.values(writers)) {
    if (writer.calledWith) {
      hadOutput = writer.calledWithMatch(matchOutput);
      if (hadOutput) break;
    }
  }

  return hadOutput;
}

function assertOutputContains(writers, output) {
  if (!someoneHadOutput(writers, output)) {
    throw new Error(`Expected something to have been called with: '${output}'`);
  }
}

function assertOutputDoesntContain(writers, output) {
  if (someoneHadOutput(writers, output)) {
    throw new Error(`Expected nothing to have been called with: '${output}'`);
  }
}require('source-map-support').install();
