// appium/build/lib/cli/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RingBuffer = exports.JSON_SPACES = void 0;
exports.errAndQuit = errAndQuit;
exports.log = log;
exports.spinWith = spinWith;

require("source-map-support/register");

var _ora = _interopRequireDefault(require("ora"));

const JSON_SPACES = 4;
exports.JSON_SPACES = JSON_SPACES;

function errAndQuit(json, msg) {
  if (json) {
    console.log(JSON.stringify({
      error: `${msg}`
    }, null, JSON_SPACES));
  } else {
    console.error(`${msg}`.red);

    if (msg.stderr) {
      console.error(`${msg.stderr}`.red);
    }
  }

  process.exit(1);
}

function log(json, msg) {
  !json && console.log(msg);
}

async function spinWith(json, msg, fn) {
  if (json) {
    return await fn();
  }

  const spinner = (0, _ora.default)(msg).start();
  let res;

  try {
    res = await fn();
    spinner.succeed();
    return res;
  } catch (err) {
    spinner.fail();
    throw err;
  }
}

class RingBuffer {
  constructor(size = 50) {
    this.size = size;
    this.buffer = [];
  }

  getBuff() {
    return this.buffer;
  }

  dequeue() {
    this.buffer.shift();
  }

  enqueue(item) {
    if (this.buffer.length >= this.size) {
      this.dequeue();
    }

    this.buffer.push(item);
  }

}

exports.RingBuffer = RingBuffer;require('source-map-support').install();
