// appium/build/commands-yml/validator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.validator = exports.default = exports.CLIENT_URL_TYPES = void 0;

require("source-map-support/register");

var _validate = _interopRequireDefault(require("validate.js"));

var _lodash = _interopRequireDefault(require("lodash"));

_validate.default.validators.array = function array(value, options, key, attributes) {
  if (attributes[key] && !_validate.default.isArray(attributes[key])) {
    return `must be an array`;
  }
};

_validate.default.validators.hasAttributes = function hasAttributes(value, options) {
  if (!value) {
    return;
  }

  if (!_lodash.default.isArray(value)) {
    value = [value];
  }

  for (const item of value) {
    for (const option of options) {
      if (_lodash.default.isUndefined(item[option])) {
        return `must have attributes: ${options}`;
      }
    }
  }
};

_validate.default.validators.hasPossibleAttributes = function hasPossibleAttributes(value, options) {
  if (!value) {
    return;
  }

  if (!_lodash.default.isArray(value)) {
    return;
  }

  for (const item of value) {
    for (const key of _lodash.default.keys(item)) {
      if (!options.includes(key)) {
        return `must not include '${key}'. Available options: ${options}`;
      }
    }
  }
};

const CLIENT_URL_TYPES = {
  url: 'hostname',
  android: 'Android',
  ios: 'iOS'
};
exports.CLIENT_URL_TYPES = CLIENT_URL_TYPES;
const validator = {
  'name': {
    presence: true
  },
  'short_description': {
    presence: true
  },
  'example_usage': {},
  'example_usage.java': {},
  'example_usage.javascript_wdio': {},
  'example_usage.javascript_wd': {},
  'example_usage.ruby': {},
  'example_usage.ruby_core': {},
  'example_usage.csharp': {},
  'description': {},
  'client_docs.java': {
    hasPossibleAttributes: _lodash.default.keys(CLIENT_URL_TYPES)
  },
  'client_docs.javascript_wdio': {
    hasPossibleAttributes: _lodash.default.keys(CLIENT_URL_TYPES)
  },
  'client_docs.javascript_wd': {
    hasPossibleAttributes: _lodash.default.keys(CLIENT_URL_TYPES)
  },
  'client_docs.ruby': {
    hasPossibleAttributes: _lodash.default.keys(CLIENT_URL_TYPES)
  },
  'client_docs.ruby_core': {
    hasPossibleAttributes: _lodash.default.keys(CLIENT_URL_TYPES)
  },
  'client_docs.csharp': {
    hasPossibleAttributes: _lodash.default.keys(CLIENT_URL_TYPES)
  },
  'endpoint': {
    presence: true
  },
  'driver_support': {
    presence: true
  },
  'endpoint.url': {
    presence: true
  },
  'endpoint.url_parameters': {
    array: true,
    hasAttributes: ['name', 'description']
  },
  'endpoint.json_parameters': {
    array: true,
    hasAttributes: ['name', 'description']
  },
  'endpoint.response': {
    hasAttributes: ['type', 'description']
  },
  'specifications': {
    presence: true
  },
  'links': {
    array: true,
    hasAttributes: ['name', 'url']
  }
};
exports.validator = validator;
var _default = validator;
exports.default = _default;require('source-map-support').install();
