// appium/build/lib/drivers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KNOWN_DRIVERS = void 0;
exports.findMatchingDriver = findMatchingDriver;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _logger = _interopRequireDefault(require("./logger"));

const KNOWN_DRIVERS = {
  uiautomator2: 'appium-uiautomator2-driver',
  xcuitest: 'appium-xcuitest-driver',
  youiengine: 'appium-youiengine-driver',
  windows: 'appium-windows-driver',
  mac: 'appium-mac-driver',
  mac2: 'appium-mac2-driver',
  espresso: 'appium-espresso-driver',
  tizen: 'appium-tizen-driver',
  flutter: 'appium-flutter-driver',
  safari: 'appium-safari-driver',
  gecko: 'appium-geckodriver'
};
exports.KNOWN_DRIVERS = KNOWN_DRIVERS;

function getDriverBySupport(drivers, matchAutomationName, matchPlatformName) {
  for (const [driverName, driverData] of _lodash.default.toPairs(drivers)) {
    const {
      automationName,
      platformNames
    } = driverData;
    const aNameMatches = automationName.toLowerCase() === matchAutomationName.toLowerCase();

    const pNameMatches = _lodash.default.includes(platformNames.map(_lodash.default.toLower), matchPlatformName.toLowerCase());

    if (aNameMatches && pNameMatches) {
      return {
        driverName,
        ...driverData
      };
    }

    if (aNameMatches) {
      throw new Error(`Driver '${driverName}' supports automationName ` + `'${automationName}', but Appium could not find ` + `support for platformName '${matchPlatformName}'. Supported ` + `platformNames are: ` + JSON.stringify(platformNames));
    }
  }

  throw new Error(`Could not find installed driver to support given caps`);
}

function findMatchingDriver(config, {
  automationName,
  platformName
}) {
  if (!_lodash.default.isString(platformName)) {
    throw new Error('You must include a platformName capability');
  }

  if (!_lodash.default.isString(automationName)) {
    throw new Error('You must include an automationName capability');
  }

  _logger.default.info(`Attempting to find matching driver for automationName ` + `'${automationName}' and platformName '${platformName}'`);

  try {
    const {
      driverName,
      mainClass,
      version
    } = getDriverBySupport(config.installedExtensions, automationName, platformName);

    _logger.default.info(`The '${driverName}' driver was installed and matched caps.`);

    _logger.default.info(`Will require it at ${config.getExtensionRequirePath(driverName)}`);

    const driver = config.require(driverName);

    if (!driver) {
      throw new Error(`MainClass ${mainClass} did not result in a driver object`);
    }

    return {
      driver,
      version,
      driverName
    };
  } catch (err) {
    const msg = `Could not find a driver for automationName ` + `'${automationName}' and platformName ${platformName}'. ` + `Have you installed a driver that supports those ` + `capabilities? Run 'appium driver list --installed' to see. ` + `(Lower-level error: ${err.message})`;
    throw new Error(msg);
  }
}require('source-map-support').install();
