// appium/build/lib/driver-config.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _extensionConfig = _interopRequireDefault(require("./extension-config"));

var _extConfigIo = require("./ext-config-io");

class DriverConfig extends _extensionConfig.default {
  constructor(appiumHome, logFn) {
    super(appiumHome, _extConfigIo.DRIVER_TYPE, logFn);
    this.knownAutomationNames = new Set();
  }

  async read() {
    this.knownAutomationNames.clear();
    return await super.read();
  }

  static getInstance(appiumHome, logFn) {
    var _DriverConfig$_instan;

    const instance = (_DriverConfig$_instan = DriverConfig._instances[appiumHome]) !== null && _DriverConfig$_instan !== void 0 ? _DriverConfig$_instan : new DriverConfig(appiumHome, logFn);
    DriverConfig._instances[appiumHome] = instance;
    return instance;
  }

  getConfigProblems(extData, extName) {
    const problems = [];
    const {
      platformNames,
      automationName
    } = extData;

    if (!_lodash.default.isArray(platformNames)) {
      problems.push({
        err: 'Missing or incorrect supported platformNames list.',
        val: platformNames
      });
    } else {
      if (_lodash.default.isEmpty(platformNames)) {
        problems.push({
          err: 'Empty platformNames list.',
          val: platformNames
        });
      } else {
        for (const pName of platformNames) {
          if (!_lodash.default.isString(pName)) {
            problems.push({
              err: 'Incorrectly formatted platformName.',
              val: pName
            });
          }
        }
      }
    }

    if (!_lodash.default.isString(automationName)) {
      problems.push({
        err: 'Missing or incorrect automationName',
        val: automationName
      });
    }

    if (this.knownAutomationNames.has(automationName)) {
      problems.push({
        err: 'Multiple drivers claim support for the same automationName',
        val: automationName
      });
    }

    this.knownAutomationNames.add(automationName);
    return problems;
  }

  extensionDesc(driverName, {
    version,
    automationName
  }) {
    return `${driverName}@${version} (automationName '${automationName}')`;
  }

}

exports.default = DriverConfig;
(0, _defineProperty2.default)(DriverConfig, "_instances", {});require('source-map-support').install();
